File: backend/app/api/endpoints/recommendations.py
"""
Recommendations API — точки роста, прогнозы продаж, рекомендация закупки.

GET /api/v1/recommendations/products            — список по всем кабинетам компании
GET /api/v1/recommendations/products/{product_id} — детальная одна карточка
"""
from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel
from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.api.deps import get_current_user
from app.db.session import get_db
from app.models import OzonAccount, Product, User
from app.services.recommendations import (
    compute_list_recommendations,
    compute_product_recommendation,
)

router = APIRouter()


class _Buyout(BaseModel):
    rate: float
    confidence: str
    sample_size: int
    delivered: int
    returned: int
    arrived_total: int
    basis: str


class _Velocity(BaseModel):
    raw_avg_daily: float
    multiplier: float
    adjusted_daily: float
    confidence: str
    days_in_stock: int
    days_out_of_stock: int
    window_days: int
    total_units_sold: int
    basis: str


class _ROI(BaseModel):
    roi_pct: float
    period_days: int
    profit_rub: float
    capital_rub: float
    confidence: str
    basis: str


class ProductRecommendation(BaseModel):
    product_id: str
    product_name: str
    offer_id: str
    ozon_sku: int
    current_price: float | None
    cost_price: float | None
    image_url: str | None = None
    current_stock: int
    in_transit_to_customer: int

    buyout: _Buyout
    velocity: _Velocity
    procurement: dict[str, Any] | None
    roi_30d: _ROI | None
    abc_class: str | None
    abc_confidence: str | None

    missing_data: list[str]
    worst_cluster: dict[str, Any] | None = None


@router.get("/products", response_model=list[ProductRecommendation])
async def list_recommendations(
    cabinet_ids: list[uuid.UUID] | None = Query(
        None,
        description="Multi-select кабинетов. Если пусто — все кабинеты компании.",
    ),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> list[ProductRecommendation]:
    """Рекомендации по всем товарам компании с поправкой на фильтр кабинетов."""
    rows = await compute_list_recommendations(
        db,
        company_id=current_user.company_id,
        cabinet_ids=cabinet_ids,
    )
    return [ProductRecommendation(**row) for row in rows]


@router.get("/products/{product_id}", response_model=ProductRecommendation)
async def get_recommendation(
    product_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> ProductRecommendation:
    """Детальная рекомендация по одному товару."""
    try:
        pid = uuid.UUID(product_id)
    except ValueError:
        raise HTTPException(status_code=400, detail="Невалидный product_id")

    # Проверяем что товар принадлежит компании юзера
    result = await db.execute(
        select(Product)
        .join(OzonAccount, OzonAccount.id == Product.ozon_account_id)
        .where(
            Product.id == pid,
            OzonAccount.company_id == current_user.company_id,
            Product.deleted_at.is_(None),
        )
    )
    product = result.scalar_one_or_none()
    if not product:
        raise HTTPException(status_code=404, detail="Товар не найден")

    rec = await compute_product_recommendation(db, product)
    return ProductRecommendation(**rec)

File: frontend/src/lib/recommendations.ts
import { useQuery } from '@tanstack/react-query'
import { api } from '@/lib/api'
import { useCabinetStore } from '@/stores/cabinet'

export interface BuyoutMetric {
  rate: number
  confidence: 'high' | 'medium' | 'low'
  sample_size: number
  delivered: number
  returned: number
  arrived_total: number
  basis: string
}

export interface VelocityMetric {
  raw_avg_daily: number
  multiplier: number
  adjusted_daily: number
  confidence: 'high' | 'medium' | 'low'
  days_in_stock: number
  days_out_of_stock: number
  window_days: number
  total_units_sold: number
  basis: string
}

export interface ProcurementMetric {
  need_reorder: boolean
  /** null/Infinity = нет продаж в окне, days_left не определён. */
  days_left: number | null
  raw_need: number
  recommended_qty: number
  order_by: string | null
  projected_stockout: string | null
  confidence: 'high' | 'medium' | 'low'
  basis: string
  warnings: string[]
  signal: 'stockout' | 'reorder_now' | 'ok'
  lead_time_days: number
  safety_stock_days: number
  moq: number
  supply_params_set: boolean
}

export interface ROIMetric {
  roi_pct: number
  period_days: number
  profit_rub: number
  capital_rub: number
  confidence: 'high' | 'medium' | 'low'
  basis: string
}

export interface WorstCluster {
  cluster: string
  free_to_sell: number
  velocity_per_day: number
  days_left: number | null
  signal: 'stockout' | 'reorder_now' | 'ok'
}

export interface ProductRecommendation {
  product_id: string
  product_name: string
  offer_id: string
  ozon_sku: number
  current_price: number | null         // Ozon `price` (зачёркнутая 33000), ТОЛЬКО UI
  marketing_price?: number | null      // Ozon `marketing_seller_price` (рабочая)
  selling_price?: number | null        // canonical: marketing_price ?? current_price
  sales_percent_fbo?: number | null    // реальная %-комиссия Ozon
  cost_price: number | null
  is_archived?: boolean
  image_url: string | null
  current_stock: number
  in_transit_to_customer: number
  buyout: BuyoutMetric
  velocity: VelocityMetric
  procurement: ProcurementMetric | null
  roi_30d: ROIMetric | null
  abc_class: string | null
  abc_confidence: 'high' | 'medium' | 'low' | null
  missing_data: string[]
  worst_cluster: WorstCluster | null
}

export function useRecommendations() {
  const { selectedCabinetIds } = useCabinetStore()
  return useQuery<ProductRecommendation[]>({
    queryKey: ['recommendations', 'products', selectedCabinetIds],
    queryFn: async () => {
      const params = new URLSearchParams()
      selectedCabinetIds.forEach((id) => params.append('cabinet_ids', id))
      const qs = params.toString() ? `?${params.toString()}` : ''
      const res = await api.get(`/recommendations/products${qs}`, { timeout: 60000 })
      return res.data
    },
    staleTime: 60_000,
  })
}
